framework/physics/physics_material/mgxs/mgxs_combined.cc

```diff
@@ -0,0 +1,193 @@
+// SPDX-FileCopyrightText: 2024 The OpenSn Authors <https://open-sn.github.io/opensn/>
+// SPDX-License-Identifier: MIT
+
+#include "framework/physics/physics_material/mgxs/mgxs.h"
+#include "framework/logging/log.h"
+
+namespace opensn
+{
+
+void
+MGXS::Initialize(std::vector<std::pair<int, double>>& combinations)
+{
+  Reset();
+
+  // Pickup all xs and make sure they are valid
+  std::vector<std::shared_ptr<MGXS>> xsecs;
+  xsecs.reserve(combinations.size());
+
+  unsigned int n_grps = 0;
+  unsigned int n_precs = 0;
+  double Nf_total = 0.0; // Total density of fissile materials
+
+  // Loop over cross sections
+  for (auto combo : combinations)
+  {
+    // Get the cross section from the stack
+    std::shared_ptr<MGXS> xs;
+    xs = GetStackItemPtr(multigroup_xs_stack, combo.first, std::string(__FUNCTION__));
+    xsecs.push_back(xs);
+
+    // Set the scaling factor
+    SetScalingFactor(combo.second);
+
+    // Increment densities
+    if (xs->IsFissionable())
+    {
+      is_fissionable_ = true;
+      Nf_total += xs->ScalingFactor();
+    }
+
+    // Define and check number of groups
+    if (xsecs.size() == 1)
+      n_grps = xs->NumGroups();
+    OpenSnLogicalErrorIf(xs->NumGroups() != n_grps,
+                         "All cross sections being combined must have the same group structure.");
```

Comment:
Yes. We should improve this check. Although, I'm really of the opinion we should deprecate this capability. It seems very fragile... and has probably never been used.

Reply:
I meant deprecate the combine routine, but I see your point about combining lists of isotopes. I won't tackle improving this check in this PR, but I do still think it's needs work since all it's checking is the number of groups in each xs object and not the actual group structure itself.

Reply:
Totally agree. Right now the only checks are on number of groups and velocities (which are only required for transients which we don't have). We do have the `e_bounds` attribute (although it should probably be made into a `std::vector<double>` of length `num_groups + 1` instead of a `std::vector<std::vector<double>>`.

Reply:
I don't want to introduce too many more changes to the internals in this PR, but I hated `e_bounds` so much that I went ahead and fixed it as you suggested. It is not currently used, so it won't affect anything downstream of the mgxs class. In an OpenSn xs file, what we now expect is a list of `num_groups + 1` boundaries:

```
GROUP_STRUCTURE_BEGIN
...
...
...
GROUP_STRUCTURE_END
```